### ISSUE TITLE
Cue commands that generate json/other files should have the option --compact

### ISSUE THREAD

**Issue by rudifa** [cue-lang/cue#2477]
**Is your feature request related to a problem? Please describe.**
While commands `def`, `export` and `eval` generate indented json files, in some applications the compact (minified) json is desirable, for space and time economy.

**Describe the solution you'd like**
`cue export data.cue --compact` should produce a compact `data.json` file.

**Describe alternatives you've considered**
I can reformat the indented json file with jq or any similar tool, but getting it on demand from cue would be more efficient.

**Additional context**
The only mention of `compact` in CUE issues and in pull requests I could find is in PR #1549, suggested as a possibility by @mpvl and commented on favourably by @rogpeppe and @felixge.

While the desired effect of `--compact` is obvious for json, for other formats discussed in #1549 probably should be 'minify', i.e. remove anything that conveys no machine readable information.

**Comment by myitcv** [comment]
Modulo naming, this seems a reasonable feature request.

**Comment by jpluscplusm** [comment]
Please include the comma-delimited, white-space-removed (comments removed?) single-line form of CUE in the set of output formats being targeted by this Issue's putative option.

**Comment by myitcv** [comment]
Updated the title to reflect that this request is not specific to JSON, but could/should be applied to all output formats.

**Comment by rudifa** [comment]
#2492 was more a convergent request than a duplicate of this one.

I limited my request to JSON because it corresponds to the immediate need of my project, but 
- it is not an urgent need: it is easy to post process the indented JSON 
- it would be more efficient if compacting could be done on demand in the serialization step 

Questions, for my CUEducation:  
- is "comma-delimited, white-space-removed single-line form of CUE" format currently supported on input to CUE?
- if yes, where is this format specified?
- or would this be a display-only format?

**Comment by jpluscplusm** [comment]
[snip]
> Questions, for my CUEducation:

Unofficial answers from an enCUEsiast, rather than someone who definitively knows what they're taking about:

> - is "comma-delimited, white-space-removed single-line form of CUE" format currently supported on input to CUE?

Yes. If you use the rudimentary `tr|tr|sed` I outlined in the duplicate Issue (after having manually removed any CUE comments), you /should/ find it transforms simple enough CUE texts into acceptable single-line equivalents. As noted, it doesn't handle multi-line strings or comments, and I'm not even sure that comments /can/ be handled in a single-line form.

I'm confident there are also other corner cases it fails to deal with.

> - if yes, where is this format specified?

This section of the language guide might be useful: https://cuelang.org/docs/references/spec/#lexical-elements

I didn't use that, however - it was more trial and error :-)

> - or would this be a display-only format?

I'm not sure a display-only form of the language exists. Personally, I'm only interested in a form that can be used as input.

**Comment by rudifa** [comment]
@jpluscplusm Thank you for the guidance, I'll follow it up (not as my first priority).

By 'display-only format' I meant something like this, for use in testing
```
    assert.Equal(`{·str: "hello world"·num: 42·flt: 3.14·"k8s.io/annotation": "secure-me"·list: ["a", "b", "c", 1, 2, 3]·obj: {·foo: "bar"·L: ["a", "b", "c", 1, 2, 3]·}·}`,
		myfmt.CompactCueVal(cueval), "-")
```

**Comment by yujunz** [comment]
Is there anyone already working on this?

**Comment by mvdan** [comment]
The new formatter that @cuematthew is developing should be able to do this internally in the near term for CUE outputs. We'll probably look at this for v0.19, once the v0.18 cycle proves that the new formatter is sound.

**Comment by mvdan** [comment]
And encoding/json/v2, to be released in Go 1.27.0, similarly comes with a Compact API we could use here: https://pkg.go.dev/encoding/json/jsontext@master#Value.Compact

**Comment by rogpeppe** [comment]
ISTM like this might be more appropriate as an filetype encoding option rather than a global command-line flag,
e.g. 
```
cue export --out json+compact: foo.cue
```

**Comment by mvdan** [comment]
@rogpeppe fully agreed; I had come to the same conclusion myself :) Also because "compact" makes sense for JSON, but probably not for other newline-based encodings like TOML or INI. YAML is a bit in between, given that it's a JSON superset, but I lean against it because a one-line YAML is basically just JSON.

**Comment by mvdan** [comment]
And of course, `cue+compact:` would also be nice here, given that e.g. we can collapse multi-line structs to a single line.

One open question in my mind is whether to exclude comments. I would imagine that we keep them, i.e. `compact` doesn't affect their inclusion, and we leave that to a separate option like https://github.com/cue-lang/cue/issues/1629.

**Comment by rogpeppe** [comment]
> One open question in my mind is whether to exclude comments

If I'm after compact output, I suspect I would probably want to lose comments too by default, but have the option to retain them.

I think we could do that with the filetypes code: if compact is enabled, it will also disable comments by default but allow them to be enabled explicitly.

**Comment by mvdan** [comment]
Agreed - as discussed, because we would probably make a comments option a filetypes option as well (it is encoding-dependent as well), and in the filetypes CUE logic, we can describe that one option changes the default value of another. So for example:

* `--out cue` includes comments
* `--out cue+compact` excludes comments
* `--out cue+compact+comments=all` includes comments again

**Comment by mvdan** [comment]
https://cue.gerrithub.io/c/cue-lang/cue/+/1240382 did this for JSON and CUE, printing a single line for both. For CUE, comments are also disabled, as agreed above. We don't have a tag to re-enable comments for now, but that seems like something we can add in the future if needed.
